streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala

```diff
@@ -115,10 +113,9 @@ class KGroupedStream[K, V](val inner: KGroupedStreamJ[K, V]) {
    * latest (rolling) aggregate for each key
    * @see `org.apache.kafka.streams.kstream.KGroupedStream#aggregate`
    */ 
-  def aggregate[VR](initializer: () => VR,
-    aggregator: (K, V, VR) => VR,
-    materialized: Materialized[K, VR, ByteArrayKeyValueStore]): KTable[K, VR] =
-    inner.aggregate(initializer.asInitializer, aggregator.asAggregator, materialized)
+  def aggregate[VR](initializer: => VR)(aggregator: (K, V, VR) => VR,
```

Comment:
This is a meta comment: why we want to separate initializer with other parameters in all the places?

Reply:
@guozhangwang - As I mentioned in my comment https://github.com/apache/kafka/pull/5019#issuecomment-389225551, this has been done to aid the Scala compiler do better type inferencing. Scala compiler does type inferencing from left to right in groups. So if you place the `initializer` in a separate group, then u get better type inferencing when specifying the `initializer` in the usage. There are a few areas where we can do this for better type inference. I have some reservations on this as I mentioned in the comment I linked earlier. May be you or @ijuma can take a call on this.

Reply:
Look at the signature of `foldLeft` in Scala:
```
def foldLeft[B](z: B)(op: (B, A) => B): B
```
The zero (initializer) is curried away from the aggregator function.

Many think this is a style choice but it's not. If we wanted to implement that in one group of parameter as:
```
def foldLeft[B](z: B, op: (B, A) => B): B
```
The type parameter `B` would not be "fixed" by the type inference before we get into the aggregator function.
With currying you actually apply the first group of parameter so that we know what `B` are we talking about and then we apply the function.

Here exactly the same happens with `aggregate`, the type inference is not able to tell you from the initializer what the types of the function will be and therefore ask you to write all the types explicitly. Forcing such a thing in Scala APIs is not very common.